#### **Análisis de Eficiencia con DEA (Data Envelopment Analysis)** ####
#### **Asignatura:** Programación para Analítica Prescriptiva y de Apoyo a la Decisión ####
#### **Alumno:** Aminadab Córdova Acosta ####

### Introducción

Este notebook implementa el modelo de Análisis Envolvente de Datos (DEA) para evaluar la eficiencia relativa de una serie de concesionarios de autos. Cada unidad de análisis (DMU) se caracteriza por múltiples insumos y productos medibles, y se utiliza un enfoque de programación lineal para identificar qué unidades operan eficientemente.

El modelo sigue la formulación lineal propuesta por Charnes, Cooper y Rhodes (1978) y transformada para ser resuelta como un problema LP.

### Objetivo

Determinar la eficiencia relativa de cada concesionario automotriz considerando múltiples factores operativos, utilizando un modelo DEA orientado al producto con normalización de insumos.

### Datos

Los insumos considerados son:

- Personal
- Espacio de sala de exhibición
- Población potencial (2 segmentos)
- Consultas del modelo Alpha y Beta

Los productos considerados son:

- Ventas del modelo Alpha
- Ventas del modelo Beta
- Utilidad anual

### Metodología

Para cada concesionario (DMU):
- Se resuelve un modelo de programación lineal que maximiza la eficiencia relativa.
- Se normalizan los insumos del DMU objetivo.
- Se comparan las combinaciones de ponderaciones de entradas y salidas que lo posicionen en la frontera eficiente.

### Resultados Esperados

- Índice de eficiencia para cada DMU.
- Identificación de DMUs eficientes (eficiencia = 1).
- Identificación de DMUs ineficientes (eficiencia < 1).
- Pesos óptimos de insumos y productos para cada caso.



In [ ]:
# -*- coding: utf-8 -*-
"""
Modelo DEA (Data Envelopment Analysis) implementado en Pyomo

Este script evalúa la eficiencia relativa de múltiples unidades de decisión (DMUs),
como en el ejemplo de análisis de eficiencia de concesionarios de autos.

Formulación:
- Maximiza una combinación ponderada de salidas
- Normaliza la combinación ponderada de entradas del DMU objetivo
- Asegura que ningún DMU tenga eficiencia mayor a 1 (restricciones de cociente)

Adaptado de Gurobi a Pyomo
"""

import pandas as pd
from pyomo.environ import *

# 1. Datos de entrada: Entradas y salidas por DMU
inattr = ['staff', 'showRoom', 'Population1', 'Population2', 'alphaEnquiries', 'betaEnquiries']
outattr = ['alphaSales', 'BetaSales', 'profit']

# === Separación del gp.multidict en tres objetos: dmus, inputs, outputs ===

# Diccionario original
data = {
    'Winchester': [{'staff': 7, 'showRoom': 8, 'Population1': 10, 'Population2': 12, 'alphaEnquiries': 8.5, 'betaEnquiries': 4}, {'alphaSales': 2, 'BetaSales': 0.6, 'profit': 1.5}],
    'Andover': [{'staff': 6, 'showRoom': 6, 'Population1': 20, 'Population2': 30, 'alphaEnquiries': 9, 'betaEnquiries': 4.5}, {'alphaSales': 2.3, 'BetaSales': 0.7, 'profit': 1.6}],
    'Basingstoke': [{'staff': 2, 'showRoom': 3, 'Population1': 40, 'Population2': 40, 'alphaEnquiries': 2, 'betaEnquiries': 1.5}, {'alphaSales': 0.8, 'BetaSales': 0.25, 'profit': 0.5}],
    'Poole': [{'staff': 14, 'showRoom': 9, 'Population1': 20, 'Population2': 25, 'alphaEnquiries': 10, 'betaEnquiries': 6}, {'alphaSales': 2.6, 'BetaSales': 0.86, 'profit': 1.9}],
    'Woking': [{'staff': 10, 'showRoom': 9, 'Population1': 10, 'Population2': 10, 'alphaEnquiries': 11, 'betaEnquiries': 5}, {'alphaSales': 2.4, 'BetaSales': 1, 'profit': 2}],
    'Newbury': [{'staff': 24, 'showRoom': 15, 'Population1': 15, 'Population2': 13, 'alphaEnquiries': 25, 'betaEnquiries': 1.9}, {'alphaSales': 8, 'BetaSales': 2.6, 'profit': 4.5}],
    'Portsmouth': [{'staff': 6, 'showRoom': 7, 'Population1': 50, 'Population2': 40, 'alphaEnquiries': 8.5, 'betaEnquiries': 3}, {'alphaSales': 2.5, 'BetaSales': 0.9, 'profit': 1.6}],
    'Alresford': [{'staff': 8, 'showRoom': 7.5, 'Population1': 5, 'Population2': 8, 'alphaEnquiries': 9, 'betaEnquiries': 4}, {'alphaSales': 2.1, 'BetaSales': 0.85, 'profit': 2}],
    'Salisbury': [{'staff': 5, 'showRoom': 5, 'Population1': 10, 'Population2': 10, 'alphaEnquiries': 5, 'betaEnquiries': 2.5}, {'alphaSales': 2, 'BetaSales': 0.65, 'profit': 0.9}],
    'Guildford': [{'staff': 8, 'showRoom': 10, 'Population1': 30, 'Population2': 35, 'alphaEnquiries': 9.5, 'betaEnquiries': 4.5}, {'alphaSales': 2.05, 'BetaSales': 0.75, 'profit': 1.7}],
    'Alton': [{'staff': 7, 'showRoom': 8, 'Population1': 7, 'Population2': 8, 'alphaEnquiries': 3, 'betaEnquiries': 2}, {'alphaSales': 1.9, 'BetaSales': 0.70, 'profit': 0.5}],
    'Weybridge': [{'staff': 5, 'showRoom': 6.5, 'Population1': 9, 'Population2': 12, 'alphaEnquiries': 8, 'betaEnquiries': 4.5}, {'alphaSales': 1.8, 'BetaSales': 0.63, 'profit': 1.4}],
    'Dorchester': [{'staff': 6, 'showRoom': 7.5, 'Population1': 10, 'Population2': 10, 'alphaEnquiries': 7.5, 'betaEnquiries': 4}, {'alphaSales': 1.5, 'BetaSales': 0.45, 'profit': 1.45}],
    'Bridport': [{'staff': 11, 'showRoom': 8, 'Population1': 8, 'Population2': 10, 'alphaEnquiries': 10, 'betaEnquiries': 6}, {'alphaSales': 2.2, 'BetaSales': 0.65, 'profit': 2.2}],
    'Weymouth': [{'staff': 4, 'showRoom': 5, 'Population1': 10, 'Population2': 10, 'alphaEnquiries': 7.5, 'betaEnquiries': 3.5}, {'alphaSales': 1.8, 'BetaSales': 0.62, 'profit': 1.6}],
    'Portland': [{'staff': 3, 'showRoom': 3.5, 'Population1': 3, 'Population2': 20, 'alphaEnquiries': 2, 'betaEnquiries': 1.5}, {'alphaSales': 0.9, 'BetaSales': 0.35, 'profit': 0.5}],
    'Chichester': [{'staff': 5, 'showRoom': 5.5, 'Population1': 8, 'Population2': 10, 'alphaEnquiries': 7, 'betaEnquiries': 3.5}, {'alphaSales': 1.2, 'BetaSales': 0.45, 'profit': 1.3}],
    'Petersfield': [{'staff': 21, 'showRoom': 12, 'Population1': 6, 'Population2': 6, 'alphaEnquiries': 15, 'betaEnquiries': 8}, {'alphaSales': 6, 'BetaSales': 0.25, 'profit': 2.9}],
    'Petworth': [{'staff': 6, 'showRoom': 5.5, 'Population1': 2, 'Population2': 2, 'alphaEnquiries': 8, 'betaEnquiries': 5}, {'alphaSales': 1.5, 'BetaSales': 0.55, 'profit': 1.55}],
    'Midhurst': [{'staff': 3, 'showRoom': 3.6, 'Population1': 3, 'Population2': 3, 'alphaEnquiries': 2.5, 'betaEnquiries': 1.5}, {'alphaSales': 0.8, 'BetaSales': 0.20, 'profit': 0.45}],
    'Reading': [{'staff': 30, 'showRoom': 29, 'Population1': 120, 'Population2': 80, 'alphaEnquiries': 35, 'betaEnquiries': 20}, {'alphaSales': 7, 'BetaSales': 2.5, 'profit': 8}],
    'Southampton': [{'staff': 25, 'showRoom': 16, 'Population1': 110, 'Population2': 80, 'alphaEnquiries': 27, 'betaEnquiries': 12}, {'alphaSales': 6.5, 'BetaSales': 3.5, 'profit': 5.4}],
    'Bournemouth': [{'staff': 19, 'showRoom': 10, 'Population1': 90, 'Population2': 22, 'alphaEnquiries': 25, 'betaEnquiries': 13}, {'alphaSales': 5.5, 'BetaSales': 3.1, 'profit': 4.5}],
    'Henley': [{'staff': 7, 'showRoom': 6, 'Population1': 5, 'Population2': 7, 'alphaEnquiries': 8.5, 'betaEnquiries': 4.5}, {'alphaSales': 1.2, 'BetaSales': 0.48, 'profit': 2}],
    'Maidenhead': [{'staff': 12, 'showRoom': 8, 'Population1': 7, 'Population2': 10, 'alphaEnquiries': 12, 'betaEnquiries': 7}, {'alphaSales': 4.5, 'BetaSales': 2, 'profit': 2.3}],
    'Fareham': [{'staff': 4, 'showRoom': 6, 'Population1': 1, 'Population2': 1, 'alphaEnquiries': 7.5, 'betaEnquiries': 3.5}, {'alphaSales': 1.1, 'BetaSales': 0.48, 'profit': 1.7}],
    'Romsey': [{'staff': 2, 'showRoom': 2.5, 'Population1': 1, 'Population2': 1, 'alphaEnquiries': 2.5, 'betaEnquiries': 1}, {'alphaSales': 0.4, 'BetaSales': 0.1, 'profit': 0.55}],
    'Ringwood': [{'staff': 2, 'showRoom': 3.5, 'Population1': 2, 'Population2': 2, 'alphaEnquiries': 1.9, 'betaEnquiries': 1.2}, {'alphaSales': 0.3, 'BetaSales': 0.09, 'profit': 0.4}]
}

# Separación
dmus = list(data.keys())
inputs = {k: v[0] for k, v in data.items()}
outputs = {k: v[1] for k, v in data.items()}

dmus = list(inputs.keys())

# 2. Función para resolver el modelo DEA con Pyomo
def solve_dea_pyomo(target_dmu, verbose=True):
    model = ConcreteModel()

    # Conjuntos
    model.I = Set(initialize=inattr)
    model.R = Set(initialize=outattr)
    model.J = Set(initialize=dmus)

    # Variables: pesos de inputs y outputs
    model.v = Var(model.I, domain=NonNegativeReals)  # input weights
    model.u = Var(model.R, domain=NonNegativeReals)  # output weights

    # Restricciones de eficiencia para todos los DMUs
    def efficiency_rule(m, j):
        return sum(outputs[j][r] * m.u[r] for r in m.R) <= sum(inputs[j][i] * m.v[i] for i in m.I)
    model.EffConstraint = Constraint(model.J, rule=efficiency_rule)

    # Normalización del DMU objetivo (denominador = 1)
    model.Normalization = Constraint(expr=sum(inputs[target_dmu][i] * model.v[i] for i in model.I) == 1)

    # Función objetivo: Maximizar salidas del DMU objetivo
    model.Obj = Objective(expr=sum(outputs[target_dmu][r] * model.u[r] for r in model.R), sense=maximize)

    # Resolver
    solver = SolverFactory('glpk')  # solver gplk
    result = solver.solve(model, tee=verbose)

    # Resultados
    efficiency = value(model.Obj)

    if verbose:
        print(f"\nEficiencia del DMU '{target_dmu}': {round(efficiency, 4)}")
        print("Pesos de Inputs:")
        for i in inattr:
            print(f"  {i}: {round(value(model.v[i]), 4)}")
        print("Pesos de Outputs:")
        for r in outattr:
            print(f"  {r}: {round(value(model.u[r]), 4)}")
        print("-" * 60)

    return efficiency

# 3. Ejecutar el análisis DEA para todos los DMUs
performance = {}
for dmu in dmus:
    performance[dmu] = solve_dea_pyomo(dmu, verbose=False)

# 4. Clasificar resultados
efficient = [k for k, v in performance.items() if v >= 0.999999]
inefficient = [k for k in dmus if k not in efficient]

# 5. Reportar resultados
print("\n=== DMUs Eficientes ===")
for k in efficient:
    print(f"{k}: {round(performance[k], 4)}")

print("\n=== DMUs Ineficientes ===")
for k in inefficient:
    print(f"{k}: {round(performance[k], 4)}")



=== DMUs Eficientes ===
Basingstoke: 1.0
Newbury: 1.0
Portsmouth: 1.0
Alresford: 1.0
Salisbury: 1.0
Alton: 1.0
Weymouth: 1.0
Portland: 1.0
Petersfield: 1.0
Southampton: 1.0
Bournemouth: 1.0
Henley: 1.0
Maidenhead: 1.0
Fareham: 1.0
Romsey: 1.0

=== DMUs Ineficientes ===
Winchester: 0.8354
Andover: 0.9174
Poole: 0.8645
Woking: 0.845
Guildford: 0.8024
Weybridge: 0.8543
Dorchester: 0.8665
Bridport: 0.9816
Chichester: 0.8254
Petworth: 0.9878
Midhurst: 0.8888
Reading: 0.984
Ringwood: 0.9075


**Conclusiones**

Se aplicó el modelo DEA en su formulación lineal, implementado en Pyomo, para evaluar la eficiencia relativa de 28 concesionarios automotrices. El análisis permitió distinguir entre unidades eficientes e ineficientes con base en el uso de múltiples insumos y la generación de productos medibles.

El estudio identificó que 15 concesionarios alcanzaron un índice de eficiencia de 1.0, lo que indica que operan sobre la frontera eficiente y hacen un uso óptimo de sus recursos. Entre ellos destacan Basingstoke, Newbury, Southampton y Maidenhead.

En contraste, 13 concesionarios resultaron ineficientes, con niveles de eficiencia inferiores a 1.0. Casos como Winchester (0.8354), Guildford (0.8024) y Woking (0.845) presentan oportunidades claras de mejora. Estos resultados sugieren que, al ajustar la combinación de recursos utilizados, estos concesionarios podrían alcanzar niveles de producción más cercanos a los líderes.

En conjunto, los resultados proporcionan una base cuantitativa sólida para identificar mejores prácticas, comparar desempeños relativos y diseñar estrategias de mejora operativa dentro de la red de concesionarios evaluada.